File: app12_kotlin_android/data/src/main/kotlin/com/kotlinguard/data/cards/ReferenceValidator.kt
package com.kotlinguard.data.cards

import com.kotlinguard.data.assets.AssetSource
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * SR-07-equivalent: owaspRefs/mitreRefs values are validated against
 * allowlists bundled as Android assets before any card row is written —
 * the same guard app09's Reference_Validator and app11's ReferenceValidator
 * provide, since every one of these values is curated content (PLAN.md
 * §0.1), never extracted from the raw YAML.
 */
class ReferenceValidator(assetSource: AssetSource) {
    private val owaspAllowlist: Set<String> = loadAllowlist(assetSource, "ref-allowlists.json", "owasp_refs")
    private val mitreAllowlist: Set<String> = loadAllowlist(assetSource, "mitre-atlas-allowlist.json", "mitre_refs")

    fun assertOwaspRefsValid(refs: List<String>, cardId: String) {
        for (ref in refs) {
            if (ref !in owaspAllowlist) throw CardDecodeError.UnknownReference(ref, "owasp_refs", cardId)
        }
    }

    fun assertMitreRefsValid(refs: List<String>, cardId: String) {
        for (ref in refs) {
            if (ref !in mitreAllowlist) throw CardDecodeError.UnknownReference(ref, "mitre_refs", cardId)
        }
    }

    private companion object {
        fun loadAllowlist(assetSource: AssetSource, fileName: String, key: String): Set<String> {
            val text = assetSource.readText(fileName) ?: return emptySet()
            val root = Json.parseToJsonElement(text).jsonObject
            return root[key]?.jsonArray?.map { it.jsonPrimitive.content }?.toSet() ?: emptySet()
        }
    }
}

File: app13_ruby_FastApi/backend/app/services/reference_validator.rb
# frozen_string_literal: true

require "json"
require_relative "card_decode_error"

# SR-07-equivalent: owasp_refs/mitre_refs values are validated against
# bundled allowlist files before any card row is written to Postgres — the
# same guard every sibling's ReferenceValidator provides, since every one of
# these values is curated content (PLAN.md §0.1), never extracted from the
# raw YAML.
class ReferenceValidator
  def initialize(seeds_root:)
    @owasp_allowlist = load_allowlist(File.join(seeds_root, "ref-allowlists.json"), "owasp_refs")
    @mitre_allowlist = load_allowlist(File.join(seeds_root, "mitre-atlas-allowlist.json"), "mitre_refs")
  end

  def assert_owasp_refs_valid!(refs, card_id)
    refs.each do |ref|
      raise CardDecodeError::UnknownReference.new(ref, "owasp_refs", card_id) unless @owasp_allowlist.include?(ref)
    end
  end

  def assert_mitre_refs_valid!(refs, card_id)
    refs.each do |ref|
      raise CardDecodeError::UnknownReference.new(ref, "mitre_refs", card_id) unless @mitre_allowlist.include?(ref)
    end
  end

  private

  def load_allowlist(path, key)
    return Set.new unless File.exist?(path)

    JSON.parse(File.read(path))[key].to_a.to_set
  end
end
